src/backend/distributed/master/master_modify_multiple_shards.c

```diff
@@ -0,0 +1,403 @@
+static void LockShardsForModify(List *shardIntervalList);
+PG_FUNCTION_INFO_V1(master_modify_multiple_shards);
+
+
+/*
+ * master_modify_multiple_shards takes in a delete query string and pushes the
+ * query to shards. It finds shards that match the criteria defined in the
+ * delete command, generates the same delete query string for each of the
+ * found shards with distributed table name replaced with the shard name and
+ * sends the queries to the workers. It uses one-phase or two-phase commit
+ * transactions depending on citus.copy_transaction_manager value.
+ */
+Datum
+master_modify_multiple_shards(PG_FUNCTION_ARGS)
+{
+	text *queryText = PG_GETARG_TEXT_P(0);
+	char *queryString = text_to_cstring(queryText);
+	List *queryTreeList = NIL;
+	Oid relationId = InvalidOid;
+	Index tableId = 1;
+	Query *modifyQuery = NULL;
+	Node *queryTreeNode;
+	List *restrictClauseList = NIL;
+	bool isTopLevel = true;
+	bool failOK = false;
+	List *shardIntervalList = NIL;
+	List *prunedShardIntervalList = NIL;
+	int32 affectedTupleCount = 0;
+
+	PreventTransactionChain(isTopLevel, "master_modify_multiple_shards");
+
+	queryTreeNode = ParseTreeNode(queryString);
+	if (IsA(queryTreeNode, DeleteStmt))
+	{
+		DeleteStmt *deleteStatement = (DeleteStmt *) queryTreeNode;
+		relationId = RangeVarGetRelid(deleteStatement->relation, NoLock, failOK);
+		EnsureTablePermissions(relationId, ACL_DELETE);
+	}
+	else if (IsA(queryTreeNode, UpdateStmt))
+	{
+		UpdateStmt *updateStatement = (UpdateStmt *) queryTreeNode;
+		relationId = RangeVarGetRelid(updateStatement->relation, NoLock, failOK);
+		EnsureTablePermissions(relationId, ACL_UPDATE);
+	}
+	else
+	{
+		ereport(ERROR, (errmsg("query \"%s\" is not a delete nor update statement",
+							   queryString)));
+	}
+
+	CheckDistributedTable(relationId);
+
+	queryTreeList = pg_analyze_and_rewrite(queryTreeNode, queryString, NULL, 0);
+	modifyQuery = (Query *) linitial(queryTreeList);
+
+	ErrorIfModifyQueryNotSupported(modifyQuery);
+
+	shardIntervalList = LoadShardIntervalList(relationId);
+	restrictClauseList = WhereClauseList(modifyQuery->jointree);
+
+	prunedShardIntervalList =
+		PruneShardList(relationId, tableId, restrictClauseList, shardIntervalList);
+
+	LockShardsForModify(prunedShardIntervalList);
+
+	affectedTupleCount = SendQueryToShards(modifyQuery, prunedShardIntervalList);
+
+	PG_RETURN_INT32(affectedTupleCount);
+}
+
+
+/*
+ * LockShardsForModify command locks the replicas of given shard. The
+ * lock logic is slightly different from LockShards function. Basically,
+ *
+ * 1. If citus.all_modifications_commutative is set to true, then all locks
+ * are acquired as ShareLock.
+ * 2. If citus.all_modifications_commutative is false, then only the shards
+ * with 2 or more replicas are locked with ExclusiveLock. Otherwise, the
+ * lock is acquired with ShareLock.
+ */
+static void
+LockShardsForModify(List *shardIntervalList)
+{
+	if (AllModificationsCommutative)
+	{
+		LockShards(shardIntervalList, ShareLock);
```

Comment:
Ah, I got a little confused by this code. I'm wondering if there is a way to simplify it.

e.g. the following would achieve more or less the same goal:

    LOCKMODE lockMode = NoLock;

    if (AllModificationsCommutative)
    {
        lockMode = ShareLock;
    }
    else if (!HasReplication(shardIntervalList)) // check if any shards have >1 replica
    {
        lockMode = ShareLock;
    }
    else
    {
        lockMode = ExclusiveLock;
    }

    LockShards(shardIntervalList, lockMode);

This takes exclusive locks slightly more aggressively than necessary, but I think the readability of locking code is more important, since it's quite critical and quite hard to get right. We can also optimize HasReplication later on if tables get a well-defined replication factor.